src/mlpack/methods/random_forest/bootstrap.hpp

```diff
@@ -18,30 +18,210 @@ namespace mlpack {
+    arma::uword colIndicesCount) :
+    intervals(intervals),
+    sampleCount(colIndicesCount),
+    colIndices(arma::linspace<arma::uvec>(0,
+      colIndicesCount - 1,
+      colIndicesCount))
+  {
+    if (intervals.n_cols != 2)
+      throw std::invalid_argument(
+        "SequentialBootstrap::SequentialBootstrap(): "
+        "intervals must be a 2xm matrix!");
+  }
+
+  template<bool UseWeights,
+    typename MatType,
+    typename LabelsType,
+    typename WeightsType>
+  void Bootstrap(const MatType& dataset,
+    const LabelsType& labels,
+    const WeightsType& weights,
+    MatType& bootstrapDataset,
+    LabelsType& bootstrapLabels,
+    WeightsType& bootstrapWeights)
+  {
+    if (labels.n_cols != intervals.n_rows)
+      throw std::invalid_argument("SequentialBootstrap::Bootstrap(): "
+        "labels n_cols and intervals n_rows differ!");
+
+    // observations are stored as columns and dimensions
+    // (number of features) as rows.
+    if (colIndices.n_rows != dataset.n_cols) {
+      throw std::invalid_argument("SequentialBootstrap::Bootstrap(): "
+        "constructed with colIndicesCount different from "
+        "dataset.n_cols!");
+    }
+
+    const arma::uvec phi(ComputeSamples());
+
+    bootstrapDataset = dataset.cols(phi);
+    bootstrapLabels = labels.cols(phi);
+    if (UseWeights)
+      bootstrapWeights = weights.rows(phi);
+  }
+
+  /**
+   * Compute the samples of the next draw.
+   *
+   * @return A list of indices referring to the observations that should
+   *         be sampled.
+   */
+  arma::uvec ComputeSamples() const
+  {
+    std::vector<arma::u64> phi;
+
+    phi.reserve(sampleCount);
+    while (phi.size() < sampleCount)
+    {
+      const arma::vec prob(ComputeNextDrawProbabilities(phi,
+        intervals, colIndices.n_elem));
+      DiscreteDistribution d(
+        DiscreteDistribution(std::vector<arma::vec>(1u, prob)));
+
+      phi.push_back(colIndices[d.Random()[0]]);
+    }
+
+    return arma::conv_to<arma::uvec>::from(phi);
+  }
+
+  /**
+   * Compute the average uniqueness of each event at any sampling time point.
+   * The average uniqueness is a measure for how isolated an event is during
+   * its lifetime from other events.
+   *
+   * @param[in] intervals Is a `2 x m` matrix, where each column has the start
+   *                      sample and one past the end sample of an interval.
+   * @param[in] from Start row to calculate the average uniqueness from.
+   * @return The average uniqueness of the events in @p indicatorMatrix.
+   */
+  static arma::vec ComputeAverageUniqueness(
+    const IndMatType& intervals,
+    arma::uword       colIndicesCount,
+    arma::uword       from)
+  {
+    arma::rowvec concurrency(colIndicesCount, arma::fill::zeros);
+
+    for (arma::uword i(0); i < concurrency.n_cols; ++i) {
+      for (arma::uword j(0); j < intervals.n_rows; ++j) {
+        concurrency(i) += intervals(j, 0) <= i && i < intervals(j, 1) ? 1 : 0;
+      }
+    }
```

Comment:
You could do this more quickly by iterating over `intervals` and then doing, e.g., `concurrency.subvec(intervals(0, j), intervals(1, j)) += 1;`.

In any case, can you implement the algorithmic strategy I mentioned in my last review's comment?  This implementation scales as O(m^2 n) where m is the number of intervals, and I really would rather not go with that as a final version.